src/Features/CSharp/Portable/CodeRefactorings/InlineMethod/CSharpInlineMethodRefactoringProvider.cs

```diff
@@ -0,0 +1,194 @@
+        protected override bool IsValidExpressionUnderStatementExpression(ExpressionSyntax expressionNode)
+        {
+            // C# Expression Statements defined in the language reference
+            // expression_statement
+            //     : statement_expression ';'
+            //     ;
+            //
+            // statement_expression
+            //     : invocation_expression
+            //     | null_conditional_invocation_expression
+            //     | object_creation_expression
+            //     | assignment
+            //     | post_increment_expression
+            //     | post_decrement_expression
+            //     | pre_increment_expression
+            //     | pre_decrement_expression
+            //     | await_expression
+            //     ;
+            var isNullConditionalInvocationExpression =
+                expressionNode is ConditionalAccessExpressionSyntax conditionalAccessExpressionSyntax
+                && conditionalAccessExpressionSyntax.WhenNotNull.IsKind(SyntaxKind.InvocationExpression);
+
+            return expressionNode.IsKind(SyntaxKind.InvocationExpression)
+                   || isNullConditionalInvocationExpression
+                   || expressionNode.IsKind(SyntaxKind.ObjectCreationExpression)
+                   || expressionNode is AssignmentExpressionSyntax
+                   || expressionNode.IsKind(SyntaxKind.PreIncrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.PreDecrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.PostIncrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.PostDecrementExpression)
+                   || expressionNode.IsKind(SyntaxKind.AwaitExpression);
+        }
+
+        protected override bool TryGetInlineSyntaxNodeAndReplacementNodeForDelegate(
+            InvocationExpressionSyntax calleeInvocationNode,
+            IMethodSymbol calleeMethodSymbol,
+            ExpressionSyntax inlineExpressionNode,
+            StatementSyntax statementContainsCallee,
+            SyntaxGenerator syntaxGenerator,
+            out SyntaxNode? inlineSyntaxNode,
+            out SyntaxNode? syntaxNodeToReplace)
+        {
+            inlineSyntaxNode = null;
+            syntaxNodeToReplace = null;
+            if (statementContainsCallee is LocalDeclarationStatementSyntax localDeclarationSyntax
+                && IsVariableInitializerInLocalDeclarationSyntax(calleeInvocationNode, localDeclarationSyntax)
+                && localDeclarationSyntax.Declaration.Type.IsVar)
+            {
+                // Example:
+                // Before:
+                // void Caller() { var x = Callee(); }
+                // Action Callee() { return () => {}; }
+                //
+                // After inline it should be
+                // void Caller() { Action x = () => {};}
+                // Action Callee() { return () => {}; }
+                // 'var' can't be used for delegate
```

Comment:
this is only if the expression you're using is a lambda.  if it's not a lambda, it's ok.  i.e.:

```
                // void Caller() { var x = Callee(); }
                // Action Callee() { return (Action)null; }
```

There are other cases beyond lambdas for where `var` isn't allowed either.  For example, if you have a method that just returns null.  i.e.:

```
                // void Caller() { var x = Callee(); }
                // string Callee() { return null; }
```

The right thing here is to not stop using var, but to *always* insert an appropriate cast.  If you do this using the right extension helper, then the simplifier will remove it.  (i.e. the same way it does for the extra parens you add).

In this case in the code you would then be generating:

```
void Caller() { var x = (Action)(() => {});
```

Reply:
note: have test for the null case if you don't have it.  other intersting cases:

```
                // void Caller() { var x = Callee(); }
                // long Callee() { return 1; }
```

This will need a `(long)` cast in the final code.

Reply:
That's an interesting case. I think it might also happens in this case
`
void Caller() { var x = Callee(); }
IApple Callee() { return new Apple();}
`

Reply:
yup.  this is why i think you should *always* have hte casts onthe result type.

Actually, you should always have the casts on the input side as well.  Consider, for example:

```c#
void Caller(int i)
{
    Callee(i);
}

void Callee(long i)
{
    Console.WriteLine(i.GetType())
}
```

If you inline this as:

```
void Caller(int i)
{
    Console.WriteLine(i.GetType())
}
```

you change semantics.  yuo need to inline it as:

```
    Console.WriteLine(((long)i).GetType())
```

as before, simplification will get rid of the cast if unnecessary.